### ISSUE TITLE
Crash when running sqrt(-1) and log(-1)

### ISSUE THREAD

**Issue by anakimluke** [godotengine/godot#9580]
**Operating system or device - Godot version:**
Linux. Godot 2.1.4 at commit [171d8a5](https://github.com/godotengine/godot/commit/171d8a501ff97d4655ba4a7317822d41474eb4bc).

**Issue description:**
Godot crashes when trying to run `sqrt(-1)` or `log(-1)`.

**Steps to reproduce:**
1. Create a new project, add a node and a script.
2. Put  `print(sqrt(-1))` or `print(log(-1))` under `_ready()`.
3. Run the scene.
4. The editor will crash.

**Additional info:**
Terminal output when trying to run the scene:
```
ERROR: operator[]: SEVERE: Index p_index out of size (size()).
   At: core/vector.h:136.
```
When I open godot, select the project and click the `edit` button, the same error appears and the program freezes.

This bug is not on 2.1.3.

**Comment by RandomShaper** [comment]
I'll check this.

**Comment by RandomShaper** [comment]
The problem is at the [GDCompiler](https://github.com/godotengine/godot/blob/159b1fc65f650df0428dbfa9d14e3fd0bcfb94d0/modules/gdscript/gd_compiler.cpp#L1478-L1482).

Since `(NaN == NaN) == false`, the `HashMap` is re-inserting the assumed-existent key.

For me different things happen because of this: either the reported behavior or just a warning print with no crash.

I'm not sure how to handle this. Options:
- not allowing at the language level constant expressions that get reduced to _NaN_s;
- adding an special case to the `HashMap`, but that may damage performance everywhere;
- etc.

**Comment by tamkcr** [comment]
alternatively sqrt and log can return a complex number variant :D (just kidding)

**Comment by RandomShaper** [comment]
@reduz, what do you think?

I probably can fix it according to your directions (with no naughty merging XD).

**Comment by hpvb** [comment]
Are you sure this is still a problem in 3.0? I'm pretty sure this behavior is fixed in 3.0 since #7815 and I can't reproduce it on master. In my tests it just prints '-nan' as expected.

**Comment by eon-s** [comment]
Anything that prevents crashes worth doing :D

**Comment by hpvb** [comment]
@akien-mga do you want this in 2.1?

**Comment by hpvb** [comment]
I can confirm this doesn't happen in 2.1.3 either, so I guess this is a different regression. Either something from 3.0 got backported without the nan for-non-math operations fix, or something was 'just' broken.

Backporting the feature from 3.0 doesn't make much sense in the latter case.

**Comment by akien-mga** [comment]
If someone is up for a git bisect, that would be cool :)

**Comment by hpvb** [comment]
I bisected the issue to 14e30c3faadd8d53f0b8005d5100b33fcae489c5 I haven't yet looked at the commit to see what the problem may be. Testing a build with just this commit reverted to ensure this really is the problem.

**Comment by hpvb** [comment]
The specific part of the diff that causes this issue is this:
```
-       zeromem(ringptr, sizeof(RingPtr) + ringptr->size);
+       memset(ringptr, 0xEA, sizeof(RingPtr) + ringptr->size);
```

I suspect that the hashmap expects that all members are zero but with this commit they are now 0xEA. Exactly why this only fails in this case isn't entirely clear to me.

**Comment by hpvb** [comment]
This is actually not the whole story, while this does 'fix' the crash doing
```gdscript
print(sqrt(-1))
print(sqrt(-1))
```
only prints -nan once, not twice. On 2.1.3 and 3.0 it does print twice. I think that this commit 'causing' the crash is a red herring.

I'll do another set of bisections to figure out what commit caused it to only print once, or crash, I think that's probably the real culprit here.

**Comment by hpvb** [comment]
I have the same problem (-nan only printed once) when building 8df5b7151f9682e851b6c1c9ff1e13a32190b532 I'm building a release_debug build now to see if the problem remains or not.

**Comment by hpvb** [comment]
ok, yeah so... at 2.1.3:

```
$ ~/src/godot/bin/godot.x11.opt.tools.64 
-nan
-nan
$ ~/src/godot/bin/godot.x11.tools.64 
ERROR: next: Condition ' !e ' is true. returned: __null
   At: core/hash_map.h:507.
-nan
```
I'll try to do a build from HEAD in release_debug too see what we get then.

**Comment by hpvb** [comment]
Given that this *does* happen in 2.1.3 I'm going to have to assume that @RandomShaper was correct. In that case backporting #7815 will probably fix this problem.

**Comment by akien-mga** [comment]
Fixed by #10490.
